### PR TITLE
mac: improve MAC documentation (Poly 1305 key reuse)

### LINKED ISSUE #12441: MAC Salt/IV/Custom parameter confusion
Hello,

while looking into other stuff, I noticed, that there are some different parameters for MACs, which are similar or just use different names. Additionally in some cases the handling of these parameters is not consistent across the algorithms.

So, to start there are three MAC parameters I'd like to talk about (from [/include/openssl/core_names.h](https://github.com/openssl/openssl/blob/master/include/openssl/core_names.h)):
- ``#define OSSL_MAC_PARAM_IV           "iv"         /* octet string */``
- ``#define OSSL_MAC_PARAM_CUSTOM       "custom"     /* utf8 string */``
- ``#define OSSL_MAC_PARAM_SALT         "salt"       /* octet string */``

``IV`` is currently only used in GMAC, ``CUSTOM`` is used in Blake2 and KMAC and ``SALT`` is only used in Blake2.
While ``CUSTOM`` is very different conceptually from the other two, it is often called "personalization" for hashes/MACs or allso "tweak" for ciphers.
So is ``CUSTOM`` deliberately chosen to be different from the usual terms? Or should it be changed?

The other question regards the use of ``IV`` and ``SALT``:
Is it deliberate, that two different parameters are in use?
From looking at the algorithms, I would conclude, that ``SALT`` is an optional value, like ``CUSTOM``.
And ``IV`` suggests that key-reuse is not safe and can only be done with a different ``IV``.
If this is the case, it should be documented somewhere.

On the other hand, there is the Poly1305 algorithm, which originally had a Nonce as a specified parameter (which must also not be reused for a given key). But the implementation uses the version from RFC7539 and extracts the Nonce from the key:
https://github.com/openssl/openssl/blob/3d5a7578e09a984c6475b1c008f5c76f850328cb/crypto/poly1305/poly1305.c#L434-L439

Following the logic above (there is no ``IV``), the Poly1305 MAC would be safe to use multiple times with the same key, but this is not the case!

From the perspective, that a program could get any algorithm from a provider and should be able to select the correct parameters and how to work with this algorithm, there should be a way to handle the above case and communicate, that Poly1305 must only be used with one-time keys.

Are there any suggestions how to handle this?
An option would be to have a gettable parameter which is only present in such cases.

### PR DESCRIPTION
**Pull request by paulidale**
Fixes #12441

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: providers/implementations/kdfs/sskdf.c [hunk 1]
```diff
@@ -165,7 +165,7 @@ static int kmac_init(EVP_MAC_CTX *ctx, const unsigned char *custom,
     if (custom == NULL)
         return 1;
 
-    params[0] = OSSL_PARAM_construct_octet_string(OSSL_MAC_PARAM_CUSTOM,
+    params[0] = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_CUSTOM,
```
Comment (slontis) [inline providers/implementations/kdfs/sskdf.c:168]: Technically this is a MAC setter.. it just sits inside a KDF.. As long as the name doesnt change I dont care..

File: include/openssl/core_names.h [hunk 2]
```diff
@@ -186,6 +186,7 @@ extern "C" {
 #define OSSL_KDF_PARAM_SECRET       "secret"    /* octet string */
 #define OSSL_KDF_PARAM_KEY          "key"       /* octet string */
 #define OSSL_KDF_PARAM_SALT         "salt"      /* octet string */
+#define OSSL_KDF_PARAM_CUSTOM       "custom"    /* utf8 string */
```
Comment (slontis) [inline include/openssl/core_names.h:189]: Maybe use #define OSSL_ALG_PARAM_CUSTOM "custom"

#define OSSL_KDF_PARAM_CUSTOM OSSL_ALG_PARAM_CUSTOM

#define OSSL_MAC_PARAM_CUSTOM OSSL_ALG_PARAM_CUSTOM

Comment (paulidale) [reply include/openssl/core_names.h]: No need, this is gone now.

File: doc/man7/EVP_MAC-Poly1305.pod [hunk 3]
```diff
@@ -49,6 +49,11 @@ Gets the MAC size.
 The "size" parameter can also be retrieved with with EVP_MAC_CTX_get_mac_size().
 The length of the "size" parameter should not exceed that of an B<unsigned int>.
 
+=head1 NOTES
+
+The OpenSSL implementation of the Poly 1305 MAC corresponds to RFC 7539.
+As noted in RFC 8439, the key should not be reused.
```
Comment (mattcaswell) [inline doc/man7/EVP_MAC-Poly1305.pod:55]: This is quite confusing. There is 7539 and 8439. If I'm reading this correctly we conform to the earlier one, but we're repeating advice from the later one?

Comment (paulidale) [reply doc/man7/EVP_MAC-Poly1305.pod:55]: The security advice is relevant to both.  I'll see if I can reword things more clearly.

Comment (paulidale) [reply doc/man7/EVP_MAC-Poly1305.pod]: Updated.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Cleared the Post 3.0.0 milestone as it was added almost surely by a mistake.

**Comment by paulidale** [comment]
Merged to master, thanks for the comments.

**Pull request closed without merging**